arch/risc-v/src/esp32c3/esp32c3_wifi_adapter.c

```diff
@@ -1339,20 +1340,23 @@ static int IRAM_ATTR wifi_is_in_isr(void)
 static void *esp_thread_semphr_get(void)
 {
   int ret;
-  int i;
   void *sem;
-  struct tcb_s *tcb = this_task();
-  struct task_group_s *group = tcb->group;
 
-  for (i = 0; i < CONFIG_SCHED_EXIT_MAX; i++)
+  if (!g_wifi_tkey_init)
     {
-      if (group->tg_exit[i].func.on == esp_thread_semphr_free)
+      ret = tls_alloc(NULL);
+      if (ret < 0)
         {
-          break;
+          wlerr("Failed to create pthread key\n");
+          return NULL;
         }
+
+      g_wifi_thread_key = ret;
+      g_wifi_tkey_init = true;
     }
 
-  if (i >= CONFIG_SCHED_EXIT_MAX)
+  sem = (void *)tls_get_value(g_wifi_thread_key);
```

Comment:
I'm not sure. I'm not too familiar with the tls semantics. I just wanted to fix the compilation error that was dependent on tg_exit which was removed. I also have no way to test this, so all comments about this are very welcome and I will fix issues if you see them.

Reply:
Semantically it seems what I did is equal to using the group structure to store the semaphore information like the old implementation does. 

```
/****************************************************************************
 * Name: tls_alloc
 *
 * Description:
 *   Allocate a group-unique TLS data index
 *
```
Which is exactly what the old implementation did with tg_exit, if I follow the old implementation correctly.

Reply:
@pussuw  it's more simpler to copy xtensa's version to riscv:
https://github.com/apache/incubator-nuttx/blob/master/arch/xtensa/src/esp32/esp32_wifi_adapter.c#L1256-L1293

Reply:
The `arch/xtensa/src/esp32/esp32_wifi_adapter.c` use pthread API that are equivalent to `tls_` API that you used. Let's ask espressif for more explanation.

Reply:
According to an explanation by @donghengqaz the exit callback needs to be registered with on_exit() so it is called when the task terminates (via exit()), the pthread API destructor is invoked only when a pthread exits.

Using TLS as storage here should be fine, as on as `ret = on_exit(esp_thread_semphr_free, sem);` is used.

Reply:
`task_tls_destruct();` is called from `group_release` exactly when task exits, that is why I'm rising this question, but let's wait for response from Espressif

Reply:
I remember disucssing with @donghengqaz some months ago about this solution.
The usage of `on_exit` instead of relying on pthread's destructor is due to NuttX implementation of `pthread_key_create` not being fully compliant to the standard.
The standard says that the keys created via `pthread_key_create` may be shared between threads, but the destructor is still thread-private. But, if I am not mistaken, on NuttX the destructor is private to a task-group. So using pthread keys did not work out in the certain scenarios.

Reply:
Ok thanks @gustavonihei , this is one of the final small details that need to be figured out before this epos of a PR can be merged, your input is very much appreciated.

Reply:
I think that `task_tls_` level API's should provide the sufficient level to fix the leaks.

Reply:
@gustavonihei can someone at Espressif try this approach in a side of this PR?

Reply:
I'll prepare a testing environment to fully understand the root issue and how the `on_exit` solution for ESP32-C3 solves it.
Then I'll check if Task Local Storage is an option.

Reply:
@pussuw @xiaoxiang781216 @pkarashchenko I think I understood the issue once again. The Wi-Fi kernel thread creates a new semaphore for every new socket connection (e.g. from `wapi`), but when allocating it on Thread-Local Storage it fails to destroy the semaphores due to the threads creating the semaphores belonging to a different Task Group from the one that originally created the Pthread Key.

Refactoring the solution to use Task Local Storage worked and the semaphores are now properly destroyed after `wapi` finishes.

The new solution for both ESP32 and ESP32-C3 is undergoing our internal CI. I'll submit it as soon as it passes.

Reply:
Nice!

Reply:
Solution submitted here: https://github.com/apache/incubator-nuttx/pull/6322

Reply:
Done